### PR TITLE
esp8266: Include upip as frozen bytecode.

### PR DESCRIPTION
**Pull request by pfalcon**
In previous release, it was available in the release build, but to make it
actually useful, it should be always at the fingertips, so include it in
development builds too.

### CODE REVIEW

**Comment by pfalcon** [comment]
@dpgeorge , I'd like to include this, but I wonder if that's too much magic for implementation (copied from unix port). It in particular depends on mpy-tool.py being able to de-duplicate its input files. I see it does some uniq'ing, but I'm not sure it actually does de-duplication completely. I'd personally rather update those files manually.

File: esp8266/Makefile [hunk 1]
```diff
@@ -148,7 +148,7 @@ DRIVERS_SRC_C = $(addprefix drivers/,\
 SRC_S = \
 	gchelper.s \
 
-FROZEN_MPY_PY_FILES := $(shell find -L $(FROZEN_MPY_DIR) -type f -name '*.py')
+FROZEN_MPY_PY_FILES := $(shell find -L $(FROZEN_MPY_DIR) -type f -name '*.py') modules/upip.py modules/upip_utarfile.py
```
Comment (dpgeorge) [inline esp8266/Makefile]: I'm not sure why it needs to explicitly name the upip source files.  FROZEN_MPY_DIR is equal to modules (and not overridable) so if upip is in modules/ then it'll be included as frozen bytecode.

**Review by dpgeorge — COMMENTED** [review]
>  It in particular depends on mpy-tool.py being able to de-duplicate its input files

mpy-tool.py can't deduplicate.  It'll include 2 copies of the same frozen file, with the same filename, but internally the 2 names point to 2 different instances of the frozen bytecode.

**Comment by pfalcon** [comment]
> It'll include 2 copies of the same frozen file, with the same filename, but internally the 2 names point to 2 different instances of the frozen bytecode.

Well, I tried to look at it for 20s, and didn't see an obvious duplication, but as that's first time I look at it, I wouldn't know what to look for at all.

File: esp8266/Makefile (hunk 1 above)
Comment (pfalcon) [inline esp8266/Makefile]: To get the dependencies right. First, at all, so upip was uncompressed, then at right time - uncompressing upip after *.mpy's were produced from modules/ doesn't help.

I assume all the above is enough to answer my question: superfluous complex automation is worse than superfluous simple duplication.

Comment (dpgeorge) [inline esp8266/Makefile]: I see.  Isn't there a function in make to uniqify a list/array?

Comment (pfalcon) [inline esp8266/Makefile]: Actually yeah, there's. So, you still want to go "automation" route?

Comment (pfalcon) [inline esp8266/Makefile]: I'm not sure I exactly understand. Note that with unix port, where there's "automation", this automation makes using user's frozen modules complicated (almost impossible?). So, please answer which choice I should make:

1. Use this patch
2. Just add files modules/upip.py, modules/upip_tarfile.py to the repo.

Note that if choice 2 is selected, I'll likely follow up with a similar change for unix port in some time. (And the only reason I submit this patch is because unix port already does similar thing.)

Comment (dpgeorge) [inline esp8266/Makefile]: 2 is simpler.  How about untaring upip in lib/upip/ and just symlinking from there to {esp8266,unix,etc}/modules/?

Comment (pfalcon) [inline esp8266/Makefile]: > 2 is simpler. How about untaring upip in lib/upip/ and just symlinking from there to {esp8266,unix,etc}/modules/?

What about building on windows, where there commonly wouldn't be symlinks support?

@stinos : Any comments on that?

And lib/upip/ somehow doesn't feel right, lib/ so far is used mostly by C code. I'd leave it in tools/ then.

Comment (pfalcon) [inline esp8266/Makefile]: > Symlinks have been supported for a while on windows now (native command 'mklink').

Yeah, I know that you could have created symlink with NT4.0 if not NT3.5. The questions is how git commonly used by people works with symlinks? How commonly used tar?

Comment (pfalcon) [inline esp8266/Makefile]: > Btw seems github now puts badges next to names in commit comments: pfalcon is 'MicroPython member' while dpgeorge and me are labelled 'contributor'. What's up with that?

Yup, and I love that very much. Both dpgeorge and myself are 'MicroPython member' for me.

Comment (pfalcon) [inline esp8266/Makefile]: @stinos : Perhaps I'm asking too complex questions, but let me try again. Userstory: a random Window users downloads a git package, and clones a repository containing symlinks. What would they get in place of symlinks? Possible answers: a) symlinks; b) copies; c) nothing; d) clone would error out; e) unknown.

Though well, the general approach here is: a change maker smelled a possible issue, and called the Windows maintainer for consult. The Windows maintainer didn't NAK the change, so it's good to go, and if a problem will pop up later, it won't be a problem of the project, just a problem of the Windows maintainer ;-).

Comment (stinos) [inline esp8266/Makefile]: > Perhaps I'm asking too complex questions

No it's ok, I just missed the point where the symlink would be in the git repo and thought it would be created in the makefile. Looking around for an answer it is basically e): should work but depends on certain conditions: https://github.com/git-for-windows/git/wiki/Symbolic-Links. 

>  just a problem of the Windows maintainer

..or the windows user in this case. Anyway it's ok for me. I cloned a test repo using ```git clone -c core.symlinks=true  https://github.com/ariofrio/symlink-test``` and that works fine and creates file symlinks. So if it ever causes problems for someone else the answer is gonna be 'micropython side is fine, your machine isn't'.

Comment (pfalcon) [inline esp8266/Makefile]: > missed the point where the symlink would be in the git repo

Yes, sorry, there was a lot assumed by context, perhaps I should have made it clear when asked you.

> ..or the windows user in this case.

Right, the maintainer can say that ;-). The point is that I, as a change author, don't want to knowlingly or unknowingly introduce a problem which may hit someone later, so ask for a consult. Thanks for actually checking it - now we're all set. (And in many cases, just ack/nak from whatever you already know is good enough.)

**Comment by pfalcon** [comment]
Ok, unix port refactored in 61d74fdef836753cbf1cf0e9ee7b8e50cb80f41d.

**Comment by pfalcon** [comment]
esp8266 handled by 1cc168d48736cca9326835eefb8b5ccdc4d63718. esp-extra-scripts branch updated.
